presto-main/src/test/java/com/facebook/presto/type/TestBlockAndPositionNullConvention.java

```diff
@@ -45,76 +35,35 @@
     @BeforeClass
     public void setUp()
     {
-        registerScalarFunction(BLOCK_AND_POSITION_CONVENTION);
+        registerParametricScalar(FunctionWithBlockAndPositionConvention.class);
     }
 
     @Test
     public void testBlockPosition()
     {
-        assertFunction("identityFunction(9876543210)", BIGINT, 9876543210L);
-        assertFunction("identityFunction(bound_long)", BIGINT, 1234L);
+        assertFunction("test_block_position(9876543210)", BIGINT, 9876543210L);
+        assertFalse(FunctionWithBlockAndPositionConvention.hitBlockPosition.get());
+
+        assertFunction("test_block_position(bound_long)", BIGINT, 1234L);
         assertTrue(FunctionWithBlockAndPositionConvention.hitBlockPosition.get());
     }
 
+    @ScalarFunction("test_block_position")
     public static class FunctionWithBlockAndPositionConvention
-            extends SqlScalarFunction
     {
         private static final AtomicBoolean hitBlockPosition = new AtomicBoolean();
-        public static final FunctionWithBlockAndPositionConvention BLOCK_AND_POSITION_CONVENTION = new FunctionWithBlockAndPositionConvention();
-
-        private static final MethodHandle METHOD_HANDLE_BLOCK_AND_POSITION = methodHandle(FunctionWithBlockAndPositionConvention.class, "getBlockPosition", Block.class, int.class);
-        private static final MethodHandle METHOD_HANDLE_NULL_ON_NULL = methodHandle(FunctionWithBlockAndPositionConvention.class, "getLong", long.class);
-
-        protected FunctionWithBlockAndPositionConvention()
-        {
-            super(new Signature("identityFunction", FunctionKind.SCALAR, BIGINT.getTypeSignature(), BIGINT.getTypeSignature()));
-        }
 
-        @Override
-        public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
-        {
-            return new ScalarFunctionImplementation(
-                    ImmutableList.of(
-                            new ScalarImplementationChoice(
-                                    false,
-                                    ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)),
-                                    METHOD_HANDLE_NULL_ON_NULL,
-                                    Optional.empty()),
-                            new ScalarImplementationChoice(
-                                    false,
-                                    ImmutableList.of(valueTypeArgumentProperty(BLOCK_AND_POSITION)),
-                                    METHOD_HANDLE_BLOCK_AND_POSITION,
-                                    Optional.empty())),
-                    isDeterministic());
-        }
-
-        public static long getBlockPosition(Block block, int position)
+        @SqlType(StandardTypes.BIGINT)
```

Comment:
Make your test contains all 3 of: generic, specialize, and exact. Ideally, there should be at least of specialized ones, and 2 exact ones.